### DISCUSSION TITLE
Server doesn't completely stop after AbortController.abort is called

### DISCUSSION THREAD

**Discussion by Thesephi** [oakserver/oak#687] [Ideas]
Hello Oakserver community,

I'd love to double check, if it's just me, or someone else can reproduce this (even with the latest `oak@17.1.3`):

```ts
// main.ts

import { Application } from "jsr:@oak/oak";

async function startOakApplication() {
  const abortCtrl = new AbortController();
  const { signal } = abortCtrl;
  const app = new Application();
  app.use((ctx) => ctx.body = "oak is cool");
  const p = app.listen({ signal });
  setTimeout(() => abortCtrl.abort(), 2000); // simulate turning off the application after 2 seconds
  await p;
  console.log("application aborted, safe to shut down the parent process"); // <== unreachable line
}

await startOakApplication();
```

```bash
deno --node-modules-dir=auto -A main.ts
error: Top-level await promise never resolved
await startOakApplication();
^
    at <anonymous> (file:///main.ts:15:1)
```

TL;DR: according to [the doc](https://github.com/oakserver/oak?tab=readme-ov-file#closing-the-server), we should be able to `abort` the Oak application, so its promise resolves, and we can do other things afterwards. Yet in the example above, the last `console.log` line is never reached.

---

If it's reproducible for others as well, I'd love to discuss / propose a patch: https://github.com/oakserver/oak/pull/685 - I tried to explain in the PR & also the code diff should speak for itself, but very happy to elaborate on / discuss any feedback / objection.

This is my 1st time attempting to contribute to this library, if I missed something please lemme know, my appreciations!

Thanks for your time & best regards 🙂

**Comment by Thesephi** [comment]
Apologies, race condition.. I saw there was another issue filed recently as well: https://github.com/oakserver/oak/issues/686

@nomocas let's follow together!

**Reply by nomocas** [reply]
Hey @Thesephi,

Yes, I have the same issue with oak v17.3.1. Thank you for your PR, it will solve it ;).
